### PR TITLE
Alluvium Consumer Interface w/WAL2JSON support

### PR DESCRIPTION
**Pull request by themissinghlink**
This revision introduces 2 abstract classes. And strings everything together with a super simple demo that you need docker/docker-compose to run. 

An Alluvium `Consumer` is an abstract class that takes logical replication events and does something with them. It could either log them to a file. It could send them to a Kinesis firehose. It could send them to Kafka, etc. The surface area is quite wide for this class, so users can extend this however they see fit just as long as they have implemented `__call__` which will be used by psycopg2's `consume_stream` (To be abstracted more heavily when we implement MySQL support. See Maxwell by Zendesk). 

Because different databases and different WAL serialization frameworks have their own ways of representing change data capture, this library introduces the `CDCEventProcessor` abstract class which expects a `post_process` function that let's you encode all of the special sauce you need in order to process your raw cdc events without writing more consumers. (NOTE: If this is designed correctly, the number of Processors >> number of Consumers). 

The rest of the code uses these two classes to get a demo for Postgres LogicalReplication working with Wal2Json. This demo supports both formatting versions seen here: https://access.crunchydata.com/documentation/wal2json/2.0/

**Demo**
First install everything by running `make install` and start your virtualenv.

I threw in a `docker-compose.yml` and a `Dockerfile` so that you can spin up a postgres DB with all of the logical replication settings configured. You can use this for your own configurations as well. Simply run `docker-compose up -d` to spin up your postgres DB with a test db setup. 

You can connect to it via psql by doing the following (password is in `docker-compose`):

```
psql --host=localhost --port=5432 --username=test -d postgres
```

In another tab, you can use the CLI Demo I have provided. Just go into `scripts` and type

```
python generate_cdc_events.py start_logger_listener -f V1
```
and a consumer should start listening. Then go to psql, create a table, and start submitting INSERT's, UPDATE's, and DELETE's and watch everything get logged in STDOUT of your script process to see the handled consumption of your change data capture events.

You can copy the demo provided by WAL2JSON which is in the examples section of [this](https://access.crunchydata.com/documentation/wal2json/2.0/) page.

**Call Outs for reviewer**

- I would like an indication if this demo works another machine and is clear/complete. If this works, I will make a README with that demo. 
- Would also like to see what changes I need to make so that this can be lifted into production code. Right now the exception boundaries are nonexistent. Would love guidance on what that would look like in an ideal situation.

### CODE REVIEW

**Comment by codehruv** [comment]
For the PR demo description, add an instruction for `. venv/bin/activate` before running the script to generate cdc events.

File: setup.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 from setuptools import find_packages, setup
 
-install_requires = []
+install_requires = ["boto3", "psycopg2"]
```
Comment (themissinghlink) [inline setup.py:3]: self review: Add a click dependency.

File: src/alluvium/consumer.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    def post_process(self):
+        return self.cdc_event
+
+
+ProcessorType = TypeVar("CDCEventProcessor", bound=CDCEventProcessor)
+
+
+class Consumer(ABC):
+    """Defines the consumer abstract class used by psycopg2's consume_stream."""
```
Comment (themissinghlink) [inline src/alluvium/consumer.py:66]: Call out for Andrew. The error boundary conditions are non existant atm. What requirements would you all have when putting this into production? Also what is your take on python3 parameter typing versus schematics?

File: tests/test_consumer.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    processor = Wal2JsonProcessorV2(cdc_event)
+    assert processor.should_keep
+    assert processor.post_process() == {"action": "I", "foo": "bar"}
+
+    cdc_event = json.dumps({"action": "B", "foo": "bar"})
+    processor = Wal2JsonProcessorV2(cdc_event)
+    assert not processor.should_keep
+
+
+def test_kinesis_consumer(mocker, kinesis_delivery_stream):
```
Comment (themissinghlink) [inline tests/test_consumer.py:41]: call out for @codehruv. Notice how I am just checking mock calls, this isn't testing anything very useful and I would argue is super brittle because these tests will break as soon as an interface changes. Your issue would make testing a lot more useful.

**Review by themissinghlink — COMMENTED** [review]
Did a self review and added line specific call outs for reviewers to get feedback and provide context.
